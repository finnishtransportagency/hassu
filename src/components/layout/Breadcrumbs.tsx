import React, { ReactElement, useCallback, useEffect, useState } from "react";
import { breadcrumb } from "@styles/Breadcrumbs.module.css";
import { useRouter } from "next/router";
import Link from "next/link";
import log from "loglevel";

export interface RouteLabels {
  [key: string]: { label: string; hideWhenNotCurrentRoute?: boolean };
}

export interface RouteMapping {
  [key: string]: { label: string; href: string };
}

interface Props {
  routeLabels: RouteLabels;
}

const defaultLabels: RouteLabels = {
  "/": { label: "Etusivu" },
  "/yllapito": { label: "Etusivu", hideWhenNotCurrentRoute: true },
  "/yllapito/perusta": { label: "Projektin perustaminen" },
  "/suunnitelma": { label: "Suunnitelmat" },
  "/suunnitelma/[...all]": { label: "Tutki suunnitelmaa" },
  "/_error": { label: "Virhe" },
};

export default function Breadcrumbs({ routeLabels }: Props): ReactElement {
  const router = useRouter();
  const [routeMapping, setRouteMapping] = useState<RouteMapping>({});
  const isYllapito = () => typeof router?.pathname === "string" && router.pathname.startsWith("/yllapito");

  const isCurrentRoute = useCallback((route: string) => router?.pathname === route, [router.pathname]);

  useEffect(() => {
    if (router.isReady) {
      const errorRoute = "/_error";
      const isOnErrorRoute = router.pathname === "/_error";
      const pathSplitted = router.asPath.split("/");
      pathSplitted.shift();
      pathSplitted.pop();
      const pathNameSplitted = router.pathname.split("/");
      pathNameSplitted.shift();
      let routes: RouteMapping = {};
      if (isOnErrorRoute) {
        routes[errorRoute] = { ...defaultLabels[errorRoute], href: "/_error" };
      } else {
        routes = pathNameSplitted.reduce<RouteMapping>((reducer, pathname, index) => {
          const pathNameJoined = "/" + pathNameSplitted.slice(0, index + 1).join("/");
          const labels = { ...defaultLabels, ...routeLabels };
          if (!labels[pathNameJoined]?.hideWhenNotCurrentRoute || isCurrentRoute(pathNameJoined)) {
            const href = "/" + pathSplitted.slice(0, index + 1).join("/");
            const label = labels[pathNameJoined]?.label || pathname;
            reducer[pathNameJoined] = { href, label };
          }
          return reducer;
        }, {});
      }
      log.log("routes", routes);
      setRouteMapping(routes);
    }
  }, [router.isReady, router.pathname, isCurrentRoute, routeLabels, router.asPath]);

  return (
    <div className="container mb-4">
      <nav>
        <ol className={`${breadcrumb}`}>
          <li>
            <Link href={isYllapito() ? "/yllapito" : "/"}>Valtion v√§ylien suunnittelu</Link>
          </li>
          {Object.entries(routeMapping).map(([route, { href, label }]) => (
            <li key={route}>
              {!isCurrentRoute(route) ? (
                <Link href={href}>
                  <a>
                    <span>{label}</span>
                  </a>
                </Link>
              ) : (
                <span className="font-bold">{label}</span>
              )}
            </li>
          ))}
        </ol>
      </nav>
    </div>
  );
}
