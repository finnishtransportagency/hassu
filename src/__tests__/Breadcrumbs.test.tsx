/**
 * @jest-environment jsdom
 */

import React from "react";
import Breadcrumbs, { RouteLabels, generateRoutes } from "@components/layout/Breadcrumbs";
import { create, act } from "react-test-renderer";
import { NextRouter, useRouter } from "next/router";

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

describe("Breadcrumbs", () => {
  it("generates breadcrumbs correctly for error page", () => {
    const router = {
      isReady: true,
      pathname: "/_error",
      asPath: "/yllapito/projekti/1234/?locale=fi",
    } as NextRouter;
    const tree = generateRoutes(router, {});
    expect(tree).toMatchSnapshot();
  });

  it("generates breadcrumbs correctly for non-mapped /tuntematon/polku/123/123 path", () => {
    const router = {
      isReady: true,
      pathname: "/tuntematon/polku/[sid]/[pid]",
      asPath: "/tuntematon/polku/123/456/?locale=fi",
    } as NextRouter;
    const tree = generateRoutes(router, {});
    expect(tree).toMatchSnapshot();
  });

  it("generates breadcrumbs correctly for /tuntematon/polku/123/123 path", () => {
    const routeLabels: RouteLabels = {
      "/tuntematon/polku/[sid]": { label: "Testi nimi" },
      "/tuntematon/polku/[sid]/[pid]": { label: "Projektin muokkaaminen" },
    };
    const router = {
      isReady: true,
      pathname: "/tuntematon/polku/[sid]/[pid]",
      asPath: "/tuntematon/polku/123/456/?locale=fi",
    } as NextRouter;
    const tree = generateRoutes(router, routeLabels);
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly on 'known' /yllapito/perusta path", () => {
    const router = {
      isReady: true,
      pathname: "/yllapito/perusta",
      asPath: "/yllapito/perusta/",
    };
    (useRouter as jest.Mock).mockReturnValue(router);
    const tree = create(<Breadcrumbs routeLabels={{}} />);
    act(() => {
      tree.update(<Breadcrumbs routeLabels={{}} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
