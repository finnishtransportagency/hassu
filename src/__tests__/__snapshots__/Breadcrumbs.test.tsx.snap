// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs generates breadcrumbs correctly for /tuntematon/polku/123/123 path 1`] = `
Object {
  "/tuntematon": Object {
    "href": "/tuntematon",
    "label": "tuntematon",
  },
  "/tuntematon/polku": Object {
    "href": "/tuntematon/polku",
    "label": "polku",
  },
  "/tuntematon/polku/[sid]": Object {
    "href": "/tuntematon/polku/123",
    "label": "Testi nimi",
  },
  "/tuntematon/polku/[sid]/[pid]": Object {
    "href": "/tuntematon/polku/123/456",
    "label": "Projektin muokkaaminen",
  },
}
`;

exports[`Breadcrumbs generates breadcrumbs correctly for error page 1`] = `
Object {
  "/_error": Object {
    "href": "/_error",
    "label": "Virhe",
  },
}
`;

exports[`Breadcrumbs generates breadcrumbs correctly for non-mapped /tuntematon/polku/123/123 path 1`] = `
Object {
  "/tuntematon": Object {
    "href": "/tuntematon",
    "label": "tuntematon",
  },
  "/tuntematon/polku": Object {
    "href": "/tuntematon/polku",
    "label": "polku",
  },
  "/tuntematon/polku/[sid]": Object {
    "href": "/tuntematon/polku/123",
    "label": "[sid]",
  },
  "/tuntematon/polku/[sid]/[pid]": Object {
    "href": "/tuntematon/polku/123/456",
    "label": "[pid]",
  },
}
`;

exports[`Breadcrumbs renders correctly on 'known' /yllapito/perusta path 1`] = `
<div
  className="container mb-4"
>
  <nav>
    <ol
      className="breadcrumb"
    >
      <li>
        <a
          href="/yllapito"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Valtion v√§ylien suunnittelu
        </a>
      </li>
      <li>
        <span
          className="font-bold"
        >
          Projektin perustaminen
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
